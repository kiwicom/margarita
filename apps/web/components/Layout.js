// @flow

import * as React from 'react';
import { View } from 'react-native';
import { StyleSheet } from '@kiwicom/universal-components';
import {
  withLayoutContext,
  type LayoutContextState,
} from '@kiwicom/margarita-device';

import Navbar from './Navbar';

type Props = {|
  +layoutReady: boolean,
  +children: React.Node | React.ChildrenArray<React.Node>,
|};
/**
 * This component should only be used by web.
 * react-navigation handles layout for mobile
 */
export const Layout = ({ layoutReady, children }: Props) => {
  return (
    <View style={styles.container}>
      {layoutReady && (
        <>
          <Navbar />
          {children}
        </>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    web: {
      overflow: 'auto',
    },
  },
});

const layoutSelect = ({ layoutReady }: LayoutContextState) => ({
  layoutReady,
});

export default withLayoutContext(layoutSelect)(Layout);
