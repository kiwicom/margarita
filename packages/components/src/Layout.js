// @flow

import * as React from 'react';
import { View, StyleSheet } from 'react-native';

import Navbar from './Navbar';

type Props = {|
  +children: React.Node | React.ChildrenArray<React.Node>,
|};
/**
 * This component should only be used by web.
 * react-navigation handles layout for mobile
 */
export default function Layout({ children }: Props) {
  return (
    <View style={styles.container}>
      <Navbar />
      {children}
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    overflow: 'hidden',
  },
});
