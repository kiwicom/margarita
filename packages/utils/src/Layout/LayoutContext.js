// @flow

import * as React from 'react';
import { Dimensions, Platform } from 'react-native';

import withContext from '../withContext/withContext';
import { getLayout } from './LayoutContextHelpers';
import { LAYOUT } from './LayoutConstants';

type Props = {|
  +children: React.Node,
|};

type State = {|
  layoutReady: boolean,
  layout: number,
|};

/**
 * NOTE: Default `layoutReady` value needs to be `false` only on web,
 * where it's switched to `true` together with `layout` value
 * initialisation inside `componentDidMount` because of Next.js SSR.
 */
const defaultState = {
  layoutReady: Platform.OS !== 'web',
  layout:
    Platform.OS === 'web'
      ? LAYOUT.largeDesktop
      : getLayout(Dimensions.get('window').width),
};

const { Provider, Consumer } = React.createContext<State>(defaultState);

export default class LayoutContextProvider extends React.Component<
  Props,
  State,
> {
  constructor() {
    super();

    this.state = {
      ...defaultState,
    };
  }

  componentDidMount() {
    const layout = getLayout(Dimensions.get('window').width);
    if (!this.state.layoutReady || this.state.layout !== layout) {
      this.setState({
        layoutReady: true,
        layout,
      });
    }

    Dimensions.addEventListener('change', this.handleDimensionsChange);
  }

  componentWillUnmount() {
    Dimensions.removeEventListener('change', this.handleDimensionsChange);
  }

  handleDimensionsChange = ({ window }: Object) => {
    const { layout } = this.state;
    const newLayout = getLayout(window.width);
    if (layout !== newLayout) {
      this.setState({
        layout: newLayout,
      });
    }
  };

  render() {
    return <Provider value={this.state}>{this.props.children}</Provider>;
  }
}

export const withLayoutContext = (select: State => Object) =>
  withContext<State>(select, Consumer);

export type LayoutContextState = State;
