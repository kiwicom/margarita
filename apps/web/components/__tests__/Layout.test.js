// @flow

import * as React from 'react';
import { render } from 'react-native-testing-library';
import { View } from 'react-native';

import { Layout } from '../Layout';

jest.mock('next/router', () => ({
  route: '/active',
}));

it('renders', () => {
  expect(
    render(
      <Layout layoutReady>
        <View />
      </Layout>,
    ),
  ).toMatchSnapshot();
});

it('renders with multiple children', () => {
  expect(
    render(
      <Layout layoutReady>
        <View />
        <View />
      </Layout>,
    ),
  ).toMatchSnapshot();
});
