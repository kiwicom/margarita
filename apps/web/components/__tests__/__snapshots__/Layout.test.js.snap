// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "space-between",
        "paddingHorizontal": 10,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 3,
        "zIndex": 100,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <Image
        accessibilityComponentType="button"
        accessibilityTraits="button"
        accessible={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        source={
          Object {
            "uri": "https://go.kiwi.com/logo.jpg",
          }
        }
        style={
          Object {
            "height": 50,
            "marginEnd": 20,
            "width": 80,
          }
        }
      />
      <View
        accessibilityComponentType="button"
        accessibilityTraits="button"
        accessible={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "marginHorizontal": 12,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "margin": 0,
              },
              Object {
                "fontSize": 12,
              },
              Object {
                "color": "#46515e",
              },
              Object {
                "fontFamily": "Roboto",
              },
              Array [
                Object {},
                false,
              ],
            ]
          }
        >
          Travel
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        accessibilityComponentType="button"
        accessibilityTraits="button"
        accessible={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "marginHorizontal": 12,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "margin": 0,
              },
              Object {
                "fontSize": 12,
              },
              Object {
                "color": "#46515e",
              },
              Object {
                "fontFamily": "Roboto",
              },
              Array [
                Object {},
                false,
              ],
            ]
          }
        >
          Manage My Booking
        </Text>
      </View>
      <View
        accessibilityComponentType="button"
        accessibilityTraits="button"
        accessible={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "marginHorizontal": 12,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "margin": 0,
              },
              Object {
                "fontSize": 12,
              },
              Object {
                "color": "#46515e",
              },
              Object {
                "fontFamily": "Roboto",
              },
              Array [
                Object {},
                false,
              ],
            ]
          }
        >
          Profile 
        </Text>
      </View>
    </View>
  </View>
  <View />
</View>
`;

exports[`renders with multiple children 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "space-between",
        "paddingHorizontal": 10,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 3,
        "zIndex": 100,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <Image
        accessibilityComponentType="button"
        accessibilityTraits="button"
        accessible={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        source={
          Object {
            "uri": "https://go.kiwi.com/logo.jpg",
          }
        }
        style={
          Object {
            "height": 50,
            "marginEnd": 20,
            "width": 80,
          }
        }
      />
      <View
        accessibilityComponentType="button"
        accessibilityTraits="button"
        accessible={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "marginHorizontal": 12,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "margin": 0,
              },
              Object {
                "fontSize": 12,
              },
              Object {
                "color": "#46515e",
              },
              Object {
                "fontFamily": "Roboto",
              },
              Array [
                Object {},
                false,
              ],
            ]
          }
        >
          Travel
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        accessibilityComponentType="button"
        accessibilityTraits="button"
        accessible={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "marginHorizontal": 12,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "margin": 0,
              },
              Object {
                "fontSize": 12,
              },
              Object {
                "color": "#46515e",
              },
              Object {
                "fontFamily": "Roboto",
              },
              Array [
                Object {},
                false,
              ],
            ]
          }
        >
          Manage My Booking
        </Text>
      </View>
      <View
        accessibilityComponentType="button"
        accessibilityTraits="button"
        accessible={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "marginHorizontal": 12,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "margin": 0,
              },
              Object {
                "fontSize": 12,
              },
              Object {
                "color": "#46515e",
              },
              Object {
                "fontFamily": "Roboto",
              },
              Array [
                Object {},
                false,
              ],
            ]
          }
        >
          Profile 
        </Text>
      </View>
    </View>
  </View>
  <View />
  <View />
</View>
`;
